tools/eqget/EQDSK.py

```diff
@@ -0,0 +1,180 @@
+            l = f.readline().split()
+            eqdsk['stitle'] = l[0]
+            eqdsk['ind1'] = int(l[1])
+            eqdsk['nx'] = int(l[2])
+            eqdsk['ny'] = int(l[3])
+
+            nr = eqdsk['nx']
+            nz = eqdsk['ny']
+
+            # Read meta data
+            meta = []
+            pattern = re.compile(REPATTERN)
+            for i in range(4):
+                l = f.readline()
+                matches = pattern.findall(l)
+                mlist = []
+                for m in matches:
+                    mlist.append(float(m))
+
+                meta.append(mlist)
+
+            values = _next_value(f)
+            r1d = lambda n : _read_1d(values, n)
+            r2d = lambda n, m : _read_2d(values, n, m)
+
+            eqdsk['fpol'] = r1d(nr)
+            eqdsk['p'] = r1d(nr)
+            eqdsk['ffprime'] = r1d(nr)
+            eqdsk['pprime'] = r1d(nr)
+            eqdsk['psi'] = r2d(nr, nz)
+            eqdsk['q'] = r1d(nr)
+
+            eqdsk['psimesh'] = np.linspace(0, 1, nr)
+            eqdsk['rhopsi'] = np.sqrt(eqdsk['psimesh'])
+
+            # Load plasma boundary and limiter
+            #leftover = []
+            eqdsk['nbbound'] = next(values)
+            eqdsk['nblim'] = next(values)
+            if eqdsk['nbbound'] == int(eqdsk['nbbound']):
+                # Load plasma boundary
+                rzplasma = r2d(2, eqdsk['nbbound'])
+                eqdsk['rplas'] = rzplasma[0,:]
+                eqdsk['zplas'] = rzplasma[1,:]
+            else:
+                eqdsk['rplas'] = np.array([])
+                eqdsk['zplas'] = np.array([])
+                #leftover.append(eqdsk['nbbound'])
+
+            if type(eqdsk['nblim']) == int:
+                # Load limiter
+                rzlimiter = r2d(2, eqdsk['nblim'])
+                eqdsk['rlim'] = rzlimiter[0,:]
+                eqdsk['zlim'] = rzlimiter[1,:]
+            else:
+                eqdsk['rlim'] = np.array([])
+                eqdsk['zlim'] = np.array([])
+                #leftover.append(eqdsk['nblim'])
+
+            # Ignore remainder of file
+
+        # Sort out meta data
+        eqdsk['rboxlen'] = meta[0][0]
+        eqdsk['zboxlen'] = meta[0][1]
+        eqdsk['rcentr'] = meta[0][2]
+
+        if eqdsk['rcentr'] < 0:
+            print(f"WARNING: R0 is negative. R0 = {eqdsk['rcentr']}")
+
+        eqdsk['rleft'] = meta[0][3]
+        eqdsk['zmid'] = meta[0][4]
+        eqdsk['raxis'] = meta[1][0]
+        eqdsk['zaxis'] = meta[1][1]
+        eqdsk['psiaxis'] = meta[1][2]
+        eqdsk['psiedge'] = meta[1][3]
+        eqdsk['bcentr'] = meta[1][4]
+        eqdsk['ip'] = meta[2][0]
+
+        eqdsk['R'] = np.linspace(eqdsk['rleft'], eqdsk['rleft']+eqdsk['rboxlen'], eqdsk['nx'])
+        eqdsk['Z'] = np.linspace(eqdsk['zmid']-eqdsk['zboxlen']/2, eqdsk['zmid']+eqdsk['zboxlen']/2, eqdsk['ny'])
+
+        # Turn psi into regular psi
+        eqdsk['psi'] = eqdsk['psi']*(eqdsk['psiedge']-eqdsk['psiaxis']) + eqdsk['psiaxis']
```

Comment:
I don't understand what this line is good for and what is the difference between "psi" and "regular psi", but it seems like it messes up the calculation on psin2d on line 371 in EqBase.py. I compared with the old version and it seems like this line caused the difference in the treatment of the psi-data, and when I removed it everything worked. It also worked fine to set override_psilim=True. Maybe the calculation of psin2d has to be adapted to be correct when turning "psi" into "regular psi"?

Reply:
This is to un-normalize "psi". The "psi" field in the EQDSK should be given in normalized units so that it varies on the interval [0,1] within the plasma. For the EqBase class, however, we want the "psi" field to have physical units, and so we recover those on this line.

But it might be possible that different codes use different definitions for this field? It is definitely needed for the EQDSK files I have.

The ``override_psilim`` parameter is more intended as a workaround parameter for poorly resolved equilibria. I ran into the issue of physically inconsistent equilibria, which had small errors in ``psiaxis``, ``psiedge`` and/or the magnetic axis location that could be resolved by automatically tuning these parameters.

Reply:
Alright, apparently the files I have already give the poloidal flux in Wb/rad. I don't see any obvious way to make the code recognize the unit for psi automatically, but maybe one should implement a flag to tell the code that the poloidal flux already have physical units?